### PR TITLE
First step toward statistics in ORC files with chunked writes

### LINKED ISSUE #5826: [BUG] ORC file-level statistics omitted with chunked writes
This came up in the discussion of https://github.com/rapidsai/cudf/issues/5707.
Currently, the file-level statistics are not written when a file is written in multiple chunks.

- [x] Figure out how to unit test this change. 
- [ ] Refactor `gather_statistic_blobs` to facilitate file stats merging across multiple chunks.
- [ ] Collect and merge file stats to write them in the final file footer.

### PR DESCRIPTION
**Pull request by hyperbolic2346**
We are making the changes necessary to fix issue #5826. This is the first of those changes and is a refactor of the way statistics are captured to better support chunked writing. The next change will include things like multiple data pointers passed to the kernels and storage between calls to write.

### CODE REVIEW

**Comment by hyperbolic2346** [comment]
I hadn't received any review comments, so I thought I would merge in my latest work here instead of making a new PR. This should be all the structure changes necessary for chunked writes. I will make another PR to enable them and persist the data necessary there.

**Review by vuule — CHANGES_REQUESTED** [review]
partial review - potential bug fix, and a few nit picks

File: cpp/src/io/orc/writer_impl.cu
Comment (hyperbolic2346) [reply cpp/src/io/orc/writer_impl.cu:1285]: It was, but since row groups are needed earlier than when we call `finish_statistics_blobs` it isn't in this structure anymore. Row group information is consumed in write still.

Comment (vuule) [reply cpp/src/io/orc/writer_impl.cu:1285]: Missed that part :( Might be good to rename `encoded_statistics`, since it only includes the ones that are written in the file footer.

File: cpp/src/io/orc/writer_impl.hpp
Comment (hyperbolic2346) [reply cpp/src/io/orc/writer_impl.hpp:340]: We can. This was created before I conceded storing off the stats_desc inside the intermediate stats. I'll update that.

File: cpp/src/io/orc/writer_impl.cu
Comment (hyperbolic2346) [reply cpp/src/io/orc/writer_impl.cu:1285]: A great idea, I'm thinking `encoded_footer_statistics`

File: cpp/src/io/orc/writer_impl.hpp
Comment (hyperbolic2346) [reply cpp/src/io/orc/writer_impl.hpp:291]: Done

File: cpp/src/io/orc/writer_impl.hpp
Comment (hyperbolic2346) [reply cpp/src/io/orc/writer_impl.hpp]: Actually we can't remove `num_stripes` because we do have the number of stripe blobs, we don't know how many stripes unless there is some math there to compute it. I've removed `num_columns` and left `num_stripes` for now.

**Comment by vuule** [comment]
rerun tests

**Comment by hyperbolic2346** [comment]
Fixed the dangling pointer issue that was tripping up python tests and refactored the code some. I think that the finish work can move back into close, but I'll leave that for the next PR so as to not complicate this one anymore.

File: cpp/src/io/orc/writer_impl.hpp
Comment (vuule) [inline cpp/src/io/orc/writer_impl.hpp:332]: just to be on the safe side (otherwise we need a ctor).
```suggestion
    int num_rows = 0;
```

File: cpp/src/io/orc/writer_impl.hpp
Comment (vuule) [inline cpp/src/io/orc/writer_impl.hpp:325]: if "empty `persisted_statistics`" implies zero rows. If it does not, we need a constructor that sets `num_rows`.
```suggestion
      col_types.clear();
      num_rows = 0;
```

**Review by vuule — APPROVED** [review]
looks good! Few very minor suggestions.

File: cpp/src/io/statistics/statistics.cuh
Comment (hyperbolic2346) [reply cpp/src/io/statistics/statistics.cuh:110]: I was torn here because the original variable inside `stats_column_desc` is named col_type. I agree though, we should change this to be consistent.

File: cpp/src/io/orc/writer_impl.cu
Comment (bdice) [inline cpp/src/io/orc/writer_impl.cu:58]: Can we re-add this blank line? It keeps the include sections separated.

File: cpp/src/io/orc/writer_impl.cu
Comment (bdice) [inline cpp/src/io/orc/writer_impl.cu:1085]: It looks like only `stats_merge_gropus[num_stat_blobs - 1]` needs to be read from this device-to-host copy. Is it possible to copy only one value (or even just the `start_chunk` and `num_chunks`) instead of the entire vector?

File: cpp/src/io/orc/writer_impl.cu
Comment (bdice) [inline cpp/src/io/orc/writer_impl.cu:1159]: The pattern `!= end_iterator` is slightly better because it doesn't require a random access iterator and doesn't require a check for "before" ordering.
```suggestion
      for (auto rg_idx_it = stripe.cbegin(); rg_idx_it != stripe.cend(); ++rg_idx_it) {
```

File: cpp/src/io/orc/writer_impl.cu
Comment (bdice) [inline cpp/src/io/orc/writer_impl.cu:1177]: Tried to improve wording here.
```suggestion
  // The rowgroup stat chunks are written out in each stripe and file-level chunks are written in
  // the footer. To prevent persisting the rowgroup stat chunks across multiple write calls in a
  // chunked write situation, these allocations are split up so stripe data can persist until the
  // footer is written.
```

File: cpp/src/io/orc/writer_impl.cu
Comment (bdice) [inline cpp/src/io/orc/writer_impl.cu:1195]: I'm not sure if I understand this line. What does "min/max strings" mean at the end? Is that an example of the type of data that must be kept until the end of the chunked write? Are statistics always stored as strings?

File: cpp/src/io/orc/writer_impl.cu
Comment (bdice) [inline cpp/src/io/orc/writer_impl.cu:1273]: It looks like this is using a raw `cudaMemcpyAsync` on a `hostdevice_vector` object. Can we use its built-in methods to avoid a raw `cudaMemcpyAsync` call?

**Review by bdice — COMMENTED** [review]
Some comments attached. I'd like another person to be the second approver for this PR since I'm not very familiar with this code and it requires a fairly intimate knowledge of the format to verify correctness.

**Comment by vuule** [comment]
> Some comments attached. I'd like another person to be the second approver for this PR since I'm not very familiar with this code and it requires a fairly intimate knowledge of the format to verify correctness.

Thank you for the review!
Well, Devavret is out, so I don't think we're getting a second reviewer who is very familiar with statistics :D FWIW, Mike went over the semantics with Devavret as well, so IMO we should be good with a non-cuIO second review.

File: cpp/src/io/orc/writer_impl.cu
Comment (hyperbolic2346) [reply cpp/src/io/orc/writer_impl.cu:1177]: I'm going to commit this and then edit it a bit, would be nice to re-review. The stripe stripe was actually different thoughts. Rowgroup stat chunks are written out in each stripe. And then stripe and file-level chunks are written in the footer. I'll start with yours and attempt to make this more clear.

File: cpp/src/io/orc/writer_impl.cu
Comment (hyperbolic2346) [reply cpp/src/io/orc/writer_impl.cu:1195]: We are saving the strings stored in the statistics chunks, which are the minimum length string and the maximum length string. I'll clean this up as well, thanks for pointing it out.

File: cpp/src/io/orc/writer_impl.cu
Comment (hyperbolic2346) [reply cpp/src/io/orc/writer_impl.cu:1273]: Nice catch!

File: cpp/src/io/orc/writer_impl.cu
Comment (hyperbolic2346) [reply cpp/src/io/orc/writer_impl.cu:58]: unintentional, I'll fix it.

File: cpp/src/io/orc/writer_impl.cu
Comment (hyperbolic2346) [reply cpp/src/io/orc/writer_impl.cu:1085]: There is a good deal of device to host and back. I kept having to reel myself in here as I didn't want to drastically alter this. There isn't a way to get a `rmm::device_buffer` element to the host and `rmm::device_uvector` requires removing the default constructor. So for now I'm going to lift the element function from `rmm::device_uvector`, but I don't like that either. I guess it is a step in the right direction...

File: cpp/src/io/orc/writer_impl.cu
Comment (hyperbolic2346) [reply cpp/src/io/orc/writer_impl.cu]: Reworded, please advise.

**Review by mythrocks — APPROVED** [review]
Sorry for the delay. It takes me a while to grok code.

I'm generally :+1: apart from some minor questions and nitpicks.

**Comment by hyperbolic2346** [comment]
@gpucibot merge

File: cpp/src/io/orc/writer_impl.hpp
Comment (mythrocks) [reply cpp/src/io/orc/writer_impl.hpp]: Argh. Sorry, this was in the old code. (These were in my review of your initial draft. I honestly thought I'd deleted this comment.)

**Pull request merged**
